scripts/extensions/videoEditor/src/VideoTimeline/VideoTimeline.tsx

```diff
@@ -0,0 +1,255 @@
+import * as React from 'react';
+import {debounce, isEmpty} from 'lodash';
+import {BarIcon} from './BarIcon';
+import {ListThumbnails} from './ListThumbnails';
+import VideoEditorContext from '../VideoEditorContext';
+import {IThumbnail} from '../interfaces';
+
+interface IProps {
+    video: React.RefObject<HTMLVideoElement>;
+    thumbnails: Array<IThumbnail>;
+    trim: {
+        start: number;
+        end: number;
+    };
+    onTrim: (start: number, end: number, runCheckIsDirty: boolean) => void;
+}
+interface IState {
+    currentTime: number;
+    thumnailsRender: Array<IThumbnail>;
+}
+
+function getStrTime(secondsTotal: number) {
+    let mins: number = Math.floor(secondsTotal / 60);
+    let secs: number = Math.floor(secondsTotal % 60);
+    let li = Math.floor((secondsTotal * 10) % 10);
+
+    return (mins < 10 ? '0' + mins : mins) + ':' + (secs < 10 ? '0' + secs : secs) + '.' + li;
+}
+
+export class VideoTimeline extends React.Component<IProps, IState> {
+    static contextType = VideoEditorContext;
+    private wrapper: React.RefObject<HTMLDivElement>;
+    private controlbar: React.RefObject<HTMLDivElement>;
+    private intervalTimer: number;
+    private PositionX: any;
+
+    constructor(props: IProps) {
+        super(props);
+        this.state = {
+            currentTime: 0,
+            thumnailsRender: [],
+        };
+        this.wrapper = React.createRef();
+        this.controlbar = React.createRef();
+        this.intervalTimer = 0;
+
+        this.handleDrag = debounce(this.handleDrag.bind(this), 5);
+        this.handleTimelineClick = this.handleTimelineClick.bind(this);
+    }
+
+    componentDidMount() {
+        // call tick every 100ms to update current time state
+        this.intervalTimer = window.setInterval(this.tick, 100);
+        document.addEventListener('dragover', this.handledragover);
```

Comment:
We use it for fixing issue `cannot drag the left-right bar of timeline in FireFox`. 
Because in Firefox, `event.clientX`(mouse position) cannot get when is dragging (drag event), it's only get in `dragover` event of window. So we use `dragover` event window to save positionX( mouse x position) and use it when is dragging.

Reply:
Got it. Can you add comments in the code when implementing such workarounds like browser APIs not working well like in this case?

You still need to remove the event listener in `componentWillUnmount` lifecycle hook.